smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/fn/FunctionRegistry.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.rulesengine.language.syntax.fn;
+
+import java.util.HashMap;
+import java.util.Optional;
+import software.amazon.smithy.rulesengine.language.stdlib.BooleanEquals;
+import software.amazon.smithy.rulesengine.language.stdlib.IsValidHostLabel;
+import software.amazon.smithy.rulesengine.language.stdlib.ParseArn;
+import software.amazon.smithy.rulesengine.language.stdlib.ParseUrl;
+import software.amazon.smithy.rulesengine.language.stdlib.PartitionFn;
+import software.amazon.smithy.rulesengine.language.stdlib.StringEquals;
+import software.amazon.smithy.rulesengine.language.stdlib.Substring;
+import software.amazon.smithy.rulesengine.language.stdlib.UriEncode;
+import software.amazon.smithy.rulesengine.language.util.LazyValue;
+
+public final class FunctionRegistry {
+    private static final LazyValue<FunctionRegistry> GLOBAL_REGISTRY =
+            LazyValue.<FunctionRegistry>builder().initializer(() -> {
+                FunctionRegistry registry = new FunctionRegistry();
+                registry.registerFunction(new StringEquals());
+                registry.registerFunction(new BooleanEquals());
+                registry.registerFunction(new PartitionFn());
+                registry.registerFunction(new IsValidHostLabel());
+                registry.registerFunction(new ParseArn());
+                registry.registerFunction(new ParseUrl());
+                registry.registerFunction(new Substring());
+                registry.registerFunction(new UriEncode());
+                return registry;
+            }).build();
+
+    private final HashMap<String, FunctionDefinition> registry = new HashMap<>();
+
+    private FunctionRegistry() {
```

Comment:
This will make it far easier to add a service discovery mechanism in the future for registering vendor functions.